ruby/phase-1.md

```diff
@@ -0,0 +1,616 @@
+worker_2.start
+
+# later
+worker_1.stop
+worker_2.stop
+```
+
+Notes:
+
+- This is not complete example of all configuration options for a worker
+- A worker will use the workflow/activity class name as a name unless it is explicitly provided (
+  `worker_2` demonstrated this approach)
+- Each worker will run in its own thread
+- A blocking `Worker#run` method will be available for convenience of running a single worker in
+  a process until SIGTERM or SIGINT is received
+- In case connection sharing is a concern we can create and memoize a connection on the instance of
+  Temporal::Configuration
+
+And here is the `Worker` interface:
+
+```ruby
+class Worker
+  # Create a new instance of a Worker
+  def self.new(
+    config: Temporal::Configuration,
+    task_queue: String, # a task queue for polling workflow and activity tasks from
+    workflows: [Workflow], # list of workflows to handle
+    activities: [Activity], # list of activity to handle
+    thread_pool: Temporal::ThreadPool # optional thread-pool for sharing between multiple workers
+  ) -> Worker
+
+  # Start the worker and block until the process exits
+  def run() -> void
+
+  # Start the worker asynchronously
+  def start() -> void
+
+  # Gracefully stop the worker (wait for all running workflow/activities to finish)
+  def stop() -> void
+end
+```
+
+Internally each worker will create a thread for task polling as well as a thread pool of a given
+size (passed to the initializer as an optional `:threads` argument). A thread pool can optionally be
+provided by the caller allowing to share the same thread pool between multiple workers. Then each
+activity task will be dispatched to an available thread (via an in-memory queue). To make sure a
+polling worker always has processing capacity a polling request will only be dispatched if there is
+an available thread in the pool (in other words the pool will have no buffer).
+
+A very similar model will be used for workflow tasks, with an added complexity of wrapping each
+workflow in its own Fiber (inside a thread). This serves two purposes:
+
+1. A workflow code can be paused and resumed during the history replay
+2. Sticky workflows can be left paused while waiting for the next workflow task
+
+Given that Fibers are very lightweight this will allow us to keep many stick workflows efficiently
+in memory and manage accordingly to the worker configuration.
+
+Here's a diagram to demonstrate the whole setup:
+
+<img src="./ruby_threads.png" width="400" align="center" alt="Temporal" />
+
+### Alternatives considered
+
+We've decided against using the `Client` to initialise the `Worker` (something that Go and Python
+SDKs are doing). It seemed like a wrong abstraction and would require making many attributes public
+on the client to make sure the worker can access them (expanding the public API of a client). Using
+a configuration object solves this and opens up further optimisation opportunities (that will be
+proposed in the 2nd phase).
```

Comment:
I mean this complexity would have to go somewhere, right? You can argue that Client should be the container for a Worker configuration, but that feels very wrong from the domain modelling point of view. Bear in mind that Client and gRPC connection are not the same.

My question is this — why do I need to construct a Client if I only need a Worker? And Worker does not depend on any of the Client's public API methods

Reply:
> why do I need to construct a Client if I only need a Worker?

Because a worker needs a client. This is the case everywhere in Temporal and always has been and Ruby will not be unique because the core worker needs the core client. That you are concerned that some stateless helpers on top of the client make it a completely separate thing from its underlying piece is unrelated IMO. You can hide it from the user, but you're only doing them a disservice of knowing where the actual underlying client/connection is held.